extensions/mssql/src/hdfs/hdfsModel.ts

```diff
@@ -78,10 +83,62 @@ export class HdfsModel {
 	 * permissions that shouldn't change need to still exist and have the same values.
 	 * @param recursive Whether to apply the changes recursively (to all sub-folders and files)
 	 */
-	public apply(recursive: boolean = false): Promise<any> {
-		// TODO Apply recursive
+	public async apply(recursive: boolean = false): Promise<void> {
+		await this.applyAclChanges(this.path);
+		if (recursive) {
+			azdata.tasks.startBackgroundOperation(
+				{
+					connection: undefined,
+					displayName: localize('mssql.recursivePermissionOpStarted', "Applying permission changes recursively under '{0}'", this.path),
+					description: '',
+					isCancelable: false,
+					operation: async op => {
+						await this.applyToChildrenRecursive(op, this.path);
+						op.updateStatus(azdata.TaskStatus.Succeeded, localize('mssql.recursivePermissionOpSucceeded', "Permission changes applied successfully."));
+					}
+				}
+			);
+		}
+	}
+
+	/**
+	 * Recursive call to apply the current set of changes to all children of this path (if any)
+	 * @param op Background operation used to track status of the task
+	 * @param path The path
+	 */
+	private async applyToChildrenRecursive(op: azdata.BackgroundOperation, path: string): Promise<void> {
+		try {
+			op.updateStatus(azdata.TaskStatus.InProgress, localize('mssql.recursivePermissionOpProgress', "Applying permission changes to '{0}'.", path));
+			const files = await this.fileSource.enumerateFiles(path, true);
+			// Apply changes to all children of this path and then recursively apply to children of any directories
+			await Promise.all(
+				[
+					files.map(file => this.applyAclChanges(file.path)),
+					files.filter(f => f.isDirectory).map(d => this.applyToChildrenRecursive(op, d.path))
```

Comment:
So does this return 1 promise for each child, each of which chains a promise.all for its children?

Reply:
Ah good catch - no this is a bug. I'd need to concat the two arrays instead. I'm going to refactor this a bit anyways so I'll just do that in another PR.